check_stability.py

```diff
@@ -487,13 +487,44 @@ class LogHandler(reader.LogHandler):
         Subclasses reader.LogHandler.
         """
         def __init__(self):
-            self.results = defaultdict(lambda: defaultdict(lambda: defaultdict(int)))
+            self.results = OrderedDict()
+
+        def find_or_create_test(self, data):
+            test_name = data["test"]
+            if self.results.get(test_name):
+                return self.results[test_name]
+
+            test = {
+                "subtests": OrderedDict(),
+                "status": defaultdict(int)
+            }
+            self.results[test_name] = test
+            return test
+
+        def find_or_create_subtest(self, data):
+            test = self.find_or_create_test(data)
+            subtest_name = data["subtest"]
+
+            if test["subtests"].get(subtest_name):
+                return test["subtests"][subtest_name]
+
+            subtest = {
+                "status": defaultdict(int),
+                "messages": set()
+            }
+            test["subtests"][subtest_name] = subtest
+
+            return subtest
 
         def test_status(self, data):
-            self.results[data["test"]][data.get("subtest")][data["status"]] += 1
+            subtest = self.find_or_create_subtest(data)
```

Comment:
I'm quite tempted to inline this since it should always exist and function calls are rather expensive. It probably doesn't matter though, so up to you.